Updated documentation on Growing.md. (#41)
mojolicious/mojo.js: docs/Growing.md

having good test coverage and code being designed for testability, which will in turn often prevent future changes from
breaking old code. Much of mojo.js was developed using TDD.

> ## Prototype

Reviewer: For consistency second level headings should be upper case.
```suggestion
## PROTOTYPE
```

Reviewer 2: Actually, the all uppercase Perl convention is not so nice for Markdown.

> Mojolicious allows you to both create a lite app and a full app. The lite app is perfect for rapid prototyping.

Reviewer: We should use mojo.js consistently rather than Mojolicious
```suggestion
mojo.js allows you to both create a lite app and a full app. The lite app is perfect for rapid prototyping.
```

You can create the lite app by:

```
  $ npx mojo create-lite-app
```

> This creates a single file.

Reviewer: Actually it will also create or update package.json.

```
index.js  // Templates can be inlined
```

`index.js` starts the mojo app server by default on port 3000 and renders a minimal webpage.

You can test simply by opening a web browser at url `localhost:3000`.

> Full Mojolicious applications on the other hand are set-up to be a well organized distribution to maximize 

Reviewer: ```suggestion
Full mojo.js applications on the other hand are set-up to be a well organized distribution to maximize 
```

maintainability.

First, create your project folder and step into it. 

```
$ mkdir my_app
$ cd my_app
```

You can now create the full app by:

```

> $ npx mojo create-full-app

Reviewer: You're missing a step here to install mojojs otherwise npx won't find mojo.
```suggestion
$ npm i @mojojs/core 
$ npx mojo create-full-app
```

```

The full app will have the following structure:

```
my_app                                // Application dir created manually
|- config.yml                         // Configuration file
|- index.js                           // Application script
|- node_modules/
|   |- *lots of node files*
|- package.json                       // Node package information and settings
|- package-lock.json                  // Describes exact tree generated in node_modules/
|- test/                              // Test directory
|   |- example.js                     // Random test
|- controllers/                       // Controller directory

> |   |- example.js                     // 

Reviewer 2: Missing comment?

|- models/                            // Model directory
|- public/                            // Static file directory (served automatically)
|   |- index.html                     // Static HTML file
|- views/                             // Views directory
|   |- example/                       // View directory for "Example"
|   |   |- welcome.html.ejs
|   |- layouts/                       // View directory for layouts

> |   |   |- default.html.ejs           // Layout view/template

Reviewer: I like these comments on the file tree.

```

Again, both application skeletons can be automatically generated with:

```
$ npx mojo create-lite-app
$ npx mojo create-full-app
```

Feature-wise both are almost equal, the only real differences are organizational, so each one can be gradually 
transformed into the other.

### Foundation

> We start our new application with a single javascript script.

Reviewer: ```suggestion
We start our new application with a single javascript file.
```
javascript script sounds a bit weird.

Reviewer 2: And written all lower case?

Reviewer: Updated my suggestion to use official casing from wikipedia.

```
$ mkdir my_app
$ cd my_app

> $ touch myapp.pl

Reviewer: ```suggestion
$ touch myapp.js
```
Not perl.

```

This will be the foundation for our login manager example application.

```
import mojo from '@mojojs/core';

const app = mojo();

> app.get('/', ctx => ctx.render({text: 'I ♥ Mojo!'}));

Reviewer: The text here should match the output in the bird's eye section below.
```suggestion
app.get('/', ctx => ctx.render({text: 'Hello world!'}));
```

app.start();
```

> The built-in development web server makes working on your application a lot of fun thanks to automatic reloading.

Reviewer: As far as I know mojo.js does not include a auto-reloader like morbo. We should recommend `nodemon` for that.
```suggestion
Now that we have a simple webapp, let's start it as a server with node
```

```

> $ node index.js server

Reviewer: ```suggestion
$ node index.js server
[640563] Web application available at http://0.0.0.0:3000/
```
We should include the output of the node command here.

```

> Just save your changes and they will be automatically in effect the next time you refresh your browser.

Reviewer: ```suggestion
Note how it outputs the process id, and the URL you can reach the server on. If you want a bit of convenience, we
recommend using [nodemon](https://github.com/remy/nodemon), which will automatically handle restarts for you.
```

### A bird's-eye view

It all starts with an HTTP request like this, sent by your browser.

```
GET / HTTP/1.1
Host: localhost:3000
```

> Once the request has been received by the web server through the event loop, it will be passed on to Mojolicious, 

Reviewer: ```suggestion
Once the request has been received by the web server through the event loop, it will be passed on to mojo.js, 
```
naming

where it will be handled in a few simple steps.

1. Check if a static file exists that would meet the requirements.

2. Try to find a route that would meet the requirements.

3. Dispatch the request to this route, usually reaching one or more actions.

4. Process the request, maybe generating a response with the renderer.

5. Return control to the web server, and if no response has been generated yet, wait for a non-blocking operation to do

> so through the event loop.

Reviewer: ```suggestion
    so through the event loop.
```
I think this looks better when aligned with it's bullet point.

With our application the router would have found an action in step 2, and rendered some text in step 4, resulting in an
HTTP response like this being sent back to the browser.

```
HTTP/1.1 200 OK

> Content-Length: 12

Reviewer: ```suggestion
Content-Type: text/plain; charset=utf-8
Content-Length: 12
```
I think maybe we should include the content-type header. (This probably also applies to original Mojolicious guide.)

Reviewer: btw, @kraih I saw our actual response here also includes a Date: header, but I'm not sure it's worth including in this example?

Reviewer 2: Considering it's not valid HTTP without a `Date` header i would include it to avoid confusion. 😁

Reviewer: Updated my suggestion to also include date.

having good test coverage and code being designed for testability, which will in turn often prevent future changes from
breaking old code. Much of mojo.js was developed using TDD.

## Prototype

Mojolicious allows you to both create a lite app and a full app. The lite app

> is perfect for rapid prototyping.

Reviewer 2: Line length we use is 120.

You can create the lite app by:

>   $ npx mojo create-lite-app          !!!!!VERIFY THIS

Reviewer 2: You forgot debug comments.

Author: Can you expand on this? I am not sure I understand. But, I am removing the !!!VERIFY THIS part.

This creates a single file.

```

> index.js  // Templates and even static files can be inlined

Reviewer 2: Static files cannot be inlined.

```

> ```index.js``` starts the mojo app server by default on port 3000 and renders a

Reviewer 2: This is bad Markdown.

minimal webpage.

You can test simply by opening a web browser at url ```localhost:3000```.

Full Mojolicious applications on the other hand are set-up to be a well 
organized distribution to maximize maintainability.

First, create your project folder and step into it. 

```bash
mkdir my_app

> cd my_app

Reviewer 2: This is inconsistently formatted.
